glean/src/server_events.rs

```diff
@@ -0,0 +1,310 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+// Required imports
+use chrono::Utc;
+use serde::{Deserialize, Serialize};
+use serde_json;
+use std::collections::HashMap;
+use uuid::Uuid;
+
+// log type string used to identify logs to process in the Moz Data Pipeline
+const GLEAN_EVENT_MOZLOG_TYPE: &str = "glean-server-event";
+
+// Code below is static, regardless of what is defined in `metrics.yaml`:
+pub struct GleanEventsLogger {
+    // Application Id to identify application per Glean standards
+    pub app_id: String,
+    // Version of application emitting the event
+    pub app_display_version: String,
+    // Channel to differentiate logs from prod/beta/staging/devel
+    pub app_channel: String,
+}
+
+// Exported type for public method parameters
+// Default impl empty values will be omitted in json from ping struct definition
+#[derive(Default, Serialize, Deserialize)]
+pub struct RequestInfo {
+    pub user_agent: String,
+    pub ip_address: String,
+}
+
+// Struct to construct the glean ping
+#[derive(Serialize, Deserialize, Debug)]
+pub struct ClientInfo {
+    telemetry_sdk_build: String,
+    first_run_date: String,
+    os: String,
+    os_version: String,
+    architecture: String,
+    app_build: String,
+    app_display_version: String,
+    app_channel: String,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct PingInfo {
+    seq: u32,
+    start_time: String,
+    end_time: String,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Ping {
+    document_namespace: String,
+    document_type: String,
+    document_version: String,
+    document_id: String,
+    user_agent: Option<String>,
+    ip_address: Option<String>,
+    payload: String,
+}
+
+type Metrics = HashMap<String, HashMap<String, serde_json::Value>>;
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct GleanEvent {
+    category: String,
+    name: String,
+    timestamp: i64,
+    extra: HashMap<String, String>,
+}
+
+pub fn new_glean_event(
+    category: &str,
+    name: &str,
+    extra: std::collections::HashMap<String, String>,
+) -> GleanEvent {
+    GleanEvent {
+        category: category.to_string(),
+        name: name.to_string(),
+        timestamp: Utc::now().timestamp_millis(),
+        extra,
+    }
+}
+#[derive(Serialize, Deserialize, Debug)]
+struct PingPayload {
+    client_info: ClientInfo,
+    ping_info: PingInfo,
+    metrics: Metrics,
+    events: Vec<GleanEvent>,
+}
+
+#[derive(Serialize, Deserialize)]
+struct LogEnvelope {
+    timestamp: String,
+    logger: String,
+    #[serde(rename = "type")]
+    log_type: String,
+    fields: Ping,
+}
+
+impl GleanEventsLogger {
+    fn create_client_info(&self) -> ClientInfo {
```

Comment:
I would strongly recommend converting this to an `impl Default` unless there are more commonly used values to initialize this struct.

Reply:
In discussing, and if ok with you, were gonna leave this one as it is closely tied to the `GleanEventsLogger` struct and needs values from the instance of said `GleanEventsLogger`, so wouldn't provide us with what we're looking for.

Reply:
Ah, because the `app_display_version` and `app_channel` are defined from the values set in `GleanEventsLogger`. 

Technically, a more rust way of doing that would be to create a
```rust
impl From<GleanEventsLogger> for ClientInfo {
   fn from(gel: GleanEventsLogger>) -> ClientInfo {
      ClientInfo {
             app_display_version: gel.app_display_version,
             app_channel: gel.app_channel, 
             ..Default::default()
      }
  }
}
```
(this is where you'd have a `impl Default for ClientInfo` that specifies the other values.)

You could then do 
```rust
let telemetry_payload: PingPayload = PingPayload {
            client_info: self.clone().into(),
```

(This also lends itself nicely to template building, because then you can just define the non-standard stuff you want to specify and trust that rust will backfill with the standard bits.)

But if this is the way that the Glean team is advising you create this, so be it.

Reply:
Sure but this also gets a bit in to the weeds of what you want for default values of `app_display_version/channel` -- I'm wondering if glean might never want those to be empty strings or "Unknown" (unlike the other fields) so you might want to discourage that from potentially ever happening by not having such a default as a footgun 🤷‍♂️

Reply:
All the other parsers (JS, Go, Ruby, Pythonm, etc) generate the string "Unknown" by default, so I think we want to keep it. The comment in one of the server implementations stated: `"Unknown" fields below are required in the Glean schema, however they are not useful in server context.` Therefore, we don't get any benefit but it is expected that the string "Unknown" is present.